### PR TITLE
#4906: global reduce sum, mean, max, min operations added

### PR DESCRIPTION
**Pull request by muthutt**
#4906: global reduce sum, mean, max, min operations added

### CODE REVIEW

File: tests/tt_eager/python_api_testing/unit_testing/test_min_max.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+)
+@pytest.mark.parametrize(
+    "layout",
+    (
+        ttl.tensor.Layout.ROW_MAJOR,
+        ttl.tensor.Layout.TILE,
+    ),
+)
+def test_min_max_for_dim_hw(shape_dim, device, kind, layout):
+    ttl.program_cache.enable()
```
Comment (TT-BrianLiu) [inline tests/tt_eager/python_api_testing/unit_testing/test_min_max.py:42]: We have a fixture for this. Is it the same as `use_program_cache`?

File: tests/tt_eager/python_api_testing/unit_testing/test_min_max.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    tt_dev = tt_npu.cpu().to(ttl.tensor.Layout.ROW_MAJOR).to_torch()
+
+    comparison_fn = torch.equal
+    if kind == "mean":
+        comparison_fn = partial(torch.isclose, atol=1e-1, rtol=1e-2)
+    assert comparison_fn(tt_dev[0, 0, 0, 0], torch.Tensor([value]).bfloat16()[0])
+
+
+def p(t):
+    print(t.cpu().to(ttl.tensor.Layout.ROW_MAJOR).to_torch())
```
Comment (TT-BrianLiu) [inline tests/tt_eager/python_api_testing/unit_testing/test_min_max.py:86]: Is this used?

File: tt_eager/tt_dnn/op_library/reduce/reduce_op.cpp [hunk 3]
```diff
@@ -267,11 +268,136 @@ Tensor sum(const Tensor &input_tensor, uint dim, const MemoryConfig& output_mem_
         return AutoFormat::format_output_tensor(output, out_shape, device, Layout::TILE);
     }
 }
+Tensor max(const Tensor &input_tensor, uint dim, const MemoryConfig& output_mem_config) {
+    TT_ASSERT( dim >= 0 && dim <= 3, "dimension have to be 0-3 only corresponding to N,C,H,W");
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/reduce/reduce_op.cpp:272]: TT_FATAL

File: tt_eager/tt_dnn/op_library/reduce/reduce_op.cpp [hunk 4]
```diff
@@ -267,11 +268,136 @@ Tensor sum(const Tensor &input_tensor, uint dim, const MemoryConfig& output_mem_
...
+        return reduce(input_tensor, ReduceOpMath::MAX, ReduceOpDim::H, scaler1, output_mem_config);
+    }
+
+    // Other sum dims will autoformat first before doing composite operations
+    Device * device;
+
+    // Get the device
+    if (input_tensor.storage_type() != StorageType::DEVICE) {
+        device = AutoFormat::GetDefaultDevice();
+        TT_ASSERT(device != nullptr, "Requires setting default device if no inputs to op are on device");
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/reduce/reduce_op.cpp:287]: TT_FATAL

File: tt_eager/tt_dnn/op_library/reduce/reduce_op.cpp [hunk 5]
```diff
@@ -267,11 +268,136 @@ Tensor sum(const Tensor &input_tensor, uint dim, const MemoryConfig& output_mem_
...
+}
+
+Tensor min(const Tensor &input_tensor, uint dim, const MemoryConfig& output_mem_config) {
+    TT_ASSERT( dim >= 0 && dim <= 3, "dimension have to be 0-3 only corresponding to N,C,H,W");
+    constexpr float scaler1 = 1.0;
+
+    if ( dim == 3 ) {
+        return reduce_min(input_tensor, ReduceOpDim::W, scaler1, output_mem_config);
+    } else if ( dim == 2 ) {
+        return reduce_min(input_tensor,ReduceOpDim::H, scaler1, output_mem_config);
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/reduce/reduce_op.cpp:334]: Spacing

File: tt_eager/tt_dnn/op_library/reduce/reduce_op.cpp [hunk 6]
```diff
@@ -267,11 +268,136 @@ Tensor sum(const Tensor &input_tensor, uint dim, const MemoryConfig& output_mem_
...
+        }
+        Tensor output = transpose(input_tensor, 0, -2, output_mem_config);
+        output = max(output, 2, output_mem_config);
+        output = transpose(output, 0, -2, output_mem_config);
+        return AutoFormat::format_output_tensor(output, out_shape, device, Layout::TILE);
+    }
+}
+
+Tensor min(const Tensor &input_tensor, uint dim, const MemoryConfig& output_mem_config) {
+    TT_ASSERT( dim >= 0 && dim <= 3, "dimension have to be 0-3 only corresponding to N,C,H,W");
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/reduce/reduce_op.cpp:328]: TT_FATAL

File: tests/tt_eager/python_api_testing/unit_testing/test_sum.py [hunk 7]
```diff
@@ -38,3 +38,36 @@ def test_softmax_for_dim_hw(shape_dim, device):
...
+@pytest.mark.parametrize(
+    "shape_dim",
+    (
+        ((1, 1, 32, 32), 3),
+        ((1, 1, 32, 32), 2),
+        ((32, 32, 32, 32), 1),
+        ((32, 32, 32, 32), 0),
+    ),  # single tile
+)
+def test_sum_global(shape_dim, device):
```
Comment (TT-BrianLiu) [inline tests/tt_eager/python_api_testing/unit_testing/test_sum.py:52]: `use_program_cache`

**Review by TT-BrianLiu — APPROVED** [review]
Please address minor comments

File: tests/tt_eager/python_api_testing/unit_testing/test_min_max.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        ((32, 32, 32, 32), 3),
+    ),
+)
+@pytest.mark.parametrize(
+    "kind",
+    (
+        # "min",
+        # "max",
+        "mean",
+    ),  # single tile
```
Comment (mywoodstock) [inline tests/tt_eager/python_api_testing/unit_testing/test_min_max.py:32]: not testing `min`, `max`?

File: tests/tt_eager/python_api_testing/unit_testing/test_sum.py [hunk 9]
```diff
@@ -1,10 +1,10 @@
-# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+# SPDX-FileCopyrightText: © 2023-24 Tenstorrent Inc.
 
 # SPDX-License-Identifier: Apache-2.0
 
 import torch
 import pytest
-
+import logging
```
Comment (mywoodstock) [inline tests/tt_eager/python_api_testing/unit_testing/test_sum.py:7]: dont we use `loguru`?

**Review by mywoodstock — APPROVED** [review]
couple of minor comments

File: tt_eager/tt_dnn/op_library/reduce/reduce_op.cpp [hunk 10]
```diff
@@ -190,7 +190,8 @@ Tensor mean_hw(const Tensor& input_tensor, const MemoryConfig& output_mem_config
     return mean(input_tensor,2,output_mem_config);
 }
 Tensor global_mean(const Tensor& input_tensor, const MemoryConfig& output_mem_config) {
-    return mean(input_tensor,4,output_mem_config);
+    float inv_volume = 1.0f/input_tensor.volume();
+    return mul_unary_sfpu( inv_volume, global_sum(input_tensor,output_mem_config), output_mem_config);
```
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/reduce/reduce_op.cpp:194]: Is there a reason we need to normalize as a separate op since reduce can take a scaler arg to do this?
Or is accuracy too low since it is prescaling in reduce compared to post scaling done here?

File: tt_eager/tt_dnn/op_library/reduce/reduce_op.cpp [hunk 11]
```diff
@@ -267,11 +268,136 @@ Tensor sum(const Tensor &input_tensor, uint dim, const MemoryConfig& output_mem_
         return AutoFormat::format_output_tensor(output, out_shape, device, Layout::TILE);
     }
 }
+Tensor max(const Tensor &input_tensor, uint dim, const MemoryConfig& output_mem_config) {
```
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/reduce/reduce_op.cpp:271]: Are there any differences between sum, min, and max besides the reduce enum to pass?

If not can this be refactored into a common function that gets called? Code maintenance on these 3 large functions is non-ideal if it's just duplicated code

**Review by tt-aho — CHANGES_REQUESTED** [review]

File: tt_eager/tt_dnn/op_library/reduce/reduce_op.cpp (hunk 11 above)
Comment (tt-aho) [reply tt_eager/tt_dnn/op_library/reduce/reduce_op.cpp:271]: To clarify I meant to keep the individual max, min, sum api endpoint, but they internally call a common function which takes the enum

File: tt_eager/tt_dnn/op_library/reduce/reduce_op.cpp (hunk 10 above)
Comment (tt-aho) [reply tt_eager/tt_dnn/op_library/reduce/reduce_op.cpp:194]: When you say `this is the only configuration of mean that works well` do you mean that accuracy is low if you use the scaler arg of reduce?

File: tt_eager/tt_dnn/op_library/reduce/reduce_op.cpp [hunk 12]
```diff
@@ -262,16 +264,47 @@ Tensor sum(const Tensor &input_tensor, uint dim, const MemoryConfig& output_mem_
...
+        output = reduce_on_dim<OpKind>(output, 2, output_mem_config);
         output = transpose(output, 0, -2, output_mem_config);
         return AutoFormat::format_output_tensor(output, out_shape, device, Layout::TILE);
     }
 }
 
-Tensor global_sum(Tensor& val, const MemoryConfig& output_mem_config) {
+
+Tensor sum(const Tensor &input_tensor, uint dim, const MemoryConfig& output_mem_config) {
+    TT_FATAL( dim >= 0 && dim <= 3, "dimension have to be 0-3 only corresponding to N,C,H,W");
```
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/reduce/reduce_op.cpp:275]: Does this assert need to be in each of these functions if it's already in `reduce_on_dim`?

**Review by tt-aho — APPROVED** [review]

**Comment by tt-aho** [comment]
I approved but I think these two comments should be addressed before merge:
logger -> loguru https://github.com/tenstorrent-metal/tt-metal/pull/4908/files#r1465659863
use_program_cache https://github.com/tenstorrent-metal/tt-metal/pull/4908/files#r1465643771

**Review by eyonland — APPROVED** [review]

**Review by arakhmati — APPROVED** [review]
